test/integration/test_timeout.py

```diff
@@ -0,0 +1,224 @@
+#  © Copyright IBM Corporation 2020.
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#       http://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import time
+import unittest
+
+import requests
+from unittest.mock import Mock
+from requests import Response
+from urllib3 import Timeout
+
+from ibmcloudant.cloudant_v1 import CloudantV1
+from ibm_cloud_sdk_core.authenticators import BasicAuthenticator, IAMAuthenticator, NoAuthAuthenticator
+from ibmcloudant.couchdb_session_authenticator import CouchDbSessionAuthenticator
+
+DEFAULT_TIMEOUT = 150  # 2.5m (=150s)
+CUSTOM_TIMEOUT = 10    # 10s
+CUSTOM_TIMEOUT_CONFIG = {'timeout': CUSTOM_TIMEOUT}
+
+class Helpers():
+
+    @staticmethod
+    def get_current_time_plus_two_minute() -> int:
+        return int(time.time()) + 120
+
+    @staticmethod
+    def get_mocked_response():
+        mock_response = Response()
+        mock_response.status_code = 200
+        mock_response._content = '{"foo":"bar"}'.encode()
+        mock_response.cookies.set("AuthSession", "bar", expires=Helpers.get_current_time_plus_two_minute())
+        return mock_response
+
+    @staticmethod
+    def mock_out_cloudant_request(srv):
+        srv.http_client.request = Mock(return_value=Helpers.get_mocked_response())
+
+    # Every test case tests an authenticator and its default timeout value at first,
+    # in the second iteration a custom overwrite possibility is checked.
+    @staticmethod
+    def defineTestCases(srv):
+        return [
+            {
+                'assert_func': Helpers.assert_default_timeout_setting,
+                'set_timeout': Helpers.set_no_timeout
+            },
+            {
+                'assert_func': Helpers.assert_custom_timeout_setting,
+                'set_timeout': srv.set_http_config
+            }
+        ]
+
+    @staticmethod
+    def set_no_timeout(new_value: dict):
+        new_value['timeout']
+
+    @staticmethod
+    def assert_default_timeout_setting(tci, call_args):
+        _, _, kwargs = call_args
+        tci.assertTrue(isinstance(kwargs['timeout'], Timeout))
```

Comment:
Not sure if this is still exposed as an instance of the `Timeout` named tuple if we switch to passing an un-named tuple - if it isn't and we can't use the `read_timeout` property I think we can comfortably assert the positional elements of the tuple - we know which one is which.

Reply:
I've changed this to check for `tuple` type: 24bad01